Source/Core/DolphinWX/LuaScriptFrame.cpp

```diff
@@ -0,0 +1,553 @@
+}
+
+// The callback function that tells the emulator what to actually press
+void LuaScriptFrame::GetValues(GCPadStatus* status)
+{
+  // We lock here to prevent an issue where pad_status could be deleted during the
+  // middle of this function's execution
+  std::unique_lock<std::mutex> lock(lua_mutex);
+
+  if (lua_thread == nullptr)
+    return;
+
+  if (pad_status->stickX != GCPadStatus::MAIN_STICK_CENTER_X)
+    status->stickX = pad_status->stickX;
+
+  if (pad_status->stickY != GCPadStatus::MAIN_STICK_CENTER_Y)
+    status->stickY = pad_status->stickY;
+
+  if (pad_status->triggerLeft != 0)
+    status->triggerLeft = pad_status->triggerLeft;
+
+  if (pad_status->triggerRight != 0)
+    status->triggerRight = pad_status->triggerRight;
+
+  if (pad_status->substickX != GCPadStatus::C_STICK_CENTER_X)
+    status->substickX = pad_status->substickX;
+
+  if (pad_status->substickY != GCPadStatus::C_STICK_CENTER_Y)
+    status->substickY = pad_status->substickY;
+
+  status->button |= pad_status->button;
+}
+
+void LuaScriptFrame::NullifyLuaThread()
+{
+  lua_thread = nullptr;
+}
+
+// Sets status to the default values
+void ClearPad(GCPadStatus* status)
+{
+  status->button = 0;
+  status->stickX = GCPadStatus::MAIN_STICK_CENTER_X;
+  status->stickY = GCPadStatus::MAIN_STICK_CENTER_Y;
+  status->triggerLeft = 0;
+  status->triggerRight = 0;
+  status->substickX = GCPadStatus::C_STICK_CENTER_X;
+  status->substickY = GCPadStatus::C_STICK_CENTER_Y;
+}
+
+// Functions to register with Lua
+#pragma region Lua_Functs
+
+// Prints a string to the text control of this frame
+int printToTextCtrl(lua_State* L)
+{
+  currentWindow->Log(lua_tostring(L, 1));
+  currentWindow->Log("\n");
+
+  return 0;
+}
+
+// Steps a frame if the emulator is paused, pauses it otherwise.
+int frameAdvance(lua_State* L)
+{
+  u64 currentFrame = Movie::GetCurrentFrame();
+  Core::DoFrameStep();
+
+  // Block until a frame has actually processed
+  // Prevents a script from executing it's main loop more than once per frame.
+  while (currentFrame == Movie::GetCurrentFrame())
+    ;
+  return 0;
+}
+
+int getFrameCount(lua_State* L)
+{
+  lua_pushinteger(L, Movie::GetCurrentFrame());
+  return 1;
+}
+
+int getMovieLength(lua_State* L)
+{
+  if (Movie::IsMovieActive())
+  {
+    lua_pushinteger(L, Movie::GetTotalFrames());
+  }
+  else
+  {
+    luaL_error(L, "No active movie.");
+  }
+  return 1;
+}
+
+int softReset(lua_State* L)
+{
+  ProcessorInterface::ResetButton_Tap();
+  return 0;
+}
+
+int saveState(lua_State* L)
+{
+  int slot = lua_tointeger(L, 1);
+
+  State::Save(slot);
+  return 0;
+}
+
+int loadState(lua_State* L)
+{
+  int slot = lua_tointeger(L, 1);
+
+  State::Load(slot);
+  return 0;
+}
+
+int getAnalog(lua_State* L)
+{
+  lua_pushinteger(L, pad_status->stickX);
+  lua_pushinteger(L, pad_status->stickY);
+
+  return 2;
+}
+
+int setAnalog(lua_State* L)
+{
+  if (lua_gettop(L) != 2)
+  {
+    luaL_error(L, "Incorrect # of parameters passed to setAnalog.\n");

```

Comment:
Should this not return then instead of continuing onwards?

It is also likely better to phrase this as:

`Incorrect number of arguments passed to setAnalog. setAnalog expects two arguments"` so that the required quantity is also known.

Reply:
Nope, luaL_error() does a longjmp.

> This function never returns, but it is an idiom to use it in C functions as return luaL_error(args).

http://www.fxcodebase.com/documents/IndicoreSDK/lua/luaL_error.html